linux/common-config: Adjust some options unavailable on i686

timschumi opened the pull request:
HID_BPF and SCHED_CLASS_EXT directly depend on BPF_JIT, so guard them with the same condition.

HAVE_RUST is explicitly only enabled for x86_64 in the x86 architecture configuration, so do not assume the kernel supports Rust when building for i686.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

timschumi committed:
linux/common-config: Guard transitive dependants on BTF_JIT

BPF_JIT is not supported on i686, so we need to make sure that options
that depend on BPF_JIT don't get enabled either.

timschumi commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -43,7 +43,7 @@ let
 
   forceRust = features.rust or false;
-  kernelSupportsRust = lib.versionAtLeast version "6.7";
+  kernelSupportsRust = lib.versionAtLeast version "6.7" && !stdenv.hostPlatform.isi686;
```
Wasn't sure where exactly this condition should go (in theory it's dependent on kernel version?) and whether I should be listing all architectures that have Rust disabled, so I put it here for now to get discussion started.

K900 replied:
https://github.com/NixOS/nixpkgs/blob/2415a6c32714d53e734385378b0cbe2a59959dc9/pkgs/os-specific/linux/kernel/generic.nix#L112-L116

timschumi replied:
I saw this location, but - other than the "known to be broken" quirk - it seemed to be reserved for conditions related to the `rustc` toolchain itself. The condition that we want to introduce here may change depending on the version number in the future (in the scheme of `(lib.versionAtLeast version "6.7" && !stdenv.hostPlatform.isi686) || lib.versionAtLeast version "6.15"` or something similar, all depends on how the selection of Rust-enabled platforms evolves).

That said, I don't have a horse in this race, so if you want to have the condition in `generic.nix` despite that then I'm happy to follow, just let me know.

K900 replied:
@alyssais @ma27 thoughts?

timschumi replied:
Currently working on a second revision / followup changes, what are we thinking about this as a concept? I'm aware that I can probably make it look a lot nicer with a few more tricks.

```nix
  kernelSupportsRust =
    (lib.versionAtLeast version "6.7" && (
      stdenv.hostPlatform.isx86_64
    ))
    || (lib.versionAtLeast version "6.8" && (
      stdenv.hostPlatform.isLoongArch64
    ))
    || (lib.versionAtLeast version "6.9" && (
      stdenv.hostPlatform.isAarch64 && stdenv.hostPlatform.isLittleEndian
    ))
    || (lib.versionAtLeast version "6.10" && (
      # GCC is known to be broken: https://lore.kernel.org/lkml/31885EDD-EF6D-4EF1-94CA-276BA7A340B7@kernel.org/T/
      # A matching restriction was officially introduced with v6.12.
      stdenv.hostPlatform.isRiscV64 && !stdenv.cc.isGNU
    ))
    || (lib.versionAtLeast version "6.13" && (
      stdenv.hostPlatform.isAarch64
    ));
```

K900 replied:
I wouldn't bother with checks this precise for kernels that are EOL anyway. We should be fine with a hard cutoff at 6.12.

Ma27 replied:
> I think we haven't been very precise here until now, and that's why it's unclear now. Probably the RISC-V condition should be part of kernelSupportsRust, and rustAvailable should be strictly whether a rust compiler is available (i.e. just the platformMatch stuff).

After looking at https://github.com/NixOS/nixpkgs/commit/d3e6c8fc751a9be31e0f68400c835ac1534aaabb I think the difference is that essentially `rustAvailable = false` can be overriden whereas `kernelSupportsRust` cannot (since it doesn't make sense to build with Rust on <6.7 since there's no Rust to build ;-) ).

The naming I chose in this commit is rather poor though, since it doesn't reflect that very well.

But that makes me wonder: do we even want to let `features.rust` override `rustAvailable` even if the build platform has no rustc that can compile to the host platform?

timschumi replied:
Finally got around to updating the PR, let me know if this is what you had in mind (or not).

timschumi commented:
ofborg or their workflow-y cousin do not seem to be able to recognize who is the maintainer for changes like this. To be perfectly honest, I am not able to do that either.

Pinging some people that look related according to recent kernel changes, apologies in advance.

@K900 @alyssais @fabianhjr

K900 commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -1054,7 +1054,7 @@ let
 
         # Enable loading HID fixups as eBPF from userspace
-        HID_BPF = whenAtLeast "6.3" yes;
+        HID_BPF = whenAtLeast "6.3" (whenPlatformHasEBPFJit yes);
```
These LGTM.

K900 commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -1281,7 +1281,7 @@ let
 
         SCHED_CORE = whenAtLeast "5.14" yes;
-        SCHED_CLASS_EXT = whenAtLeast "6.12" yes;
+        SCHED_CLASS_EXT = whenAtLeast "6.12" (whenPlatformHasEBPFJit yes);
```
These LGTM.

timschumi committed:
linux/common-config: Split Rust toolchain and kernel-based conditions

The ongoing RISC-V breakage may become version independent in the
future, move it to the correct condition before updating it with the
full list of supported architectures.

While at it, simplify the remaining rustc-specific conditions into an
`availableOn`.

timschumi committed:
linux/common-config: Spell out the current set of supported Rust targets

This keeps us from accidentally running into platforms where Rust is not
supported by the Linux kernel itself (for that particular platform).

K900 commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -43,14 +43,22 @@ let
+      stdenv.hostPlatform.isx86_64
+      || stdenv.hostPlatform.isLoongArch64
+      || stdenv.hostPlatform.isAarch64
+      || (stdenv.hostPlatform.isRiscV64 && !stdenv.cc.isGNU)
```
Would be nice to recheck if this still applies maybe

timschumi replied:
It does, it even got restricted to clang (i.e. non-GCC) builds upstream now.

K900 replied:
This is giga nitpicky, but if you have a link to where this is defined upstream, it would be very cool to link to that here.

K900 replied:
I was more thinking of https://web.git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/rust/arch-support.rst really.

timschumi replied:
fwiw, that table apparently not properly updated anyway, it still restricts ARM64 to little-endian only, even though that restriction has been gone since 6.12.

I'll try and get around to updating the link in the afternoon.

timschumi replied:
Sadly no available shortlink for this.

K900 commented on pkgs/os-specific/linux/kernel/generic.nix:
```diff
@@ -111,9 +111,7 @@ let
-      && lib.all (p: !lib.meta.platformMatch stdenv.hostPlatform p) rustc.badTargetPlatforms
-      # Known to be broken: https://lore.kernel.org/lkml/31885EDD-EF6D-4EF1-94CA-276BA7A340B7@kernel.org/T/
-      && !(stdenv.hostPlatform.isRiscV && stdenv.cc.isGNU);
+      && lib.all (p: !lib.meta.platformMatch stdenv.hostPlatform p) rustc.badTargetPlatforms;
```
This is just `availableOn` now, right?

timschumi replied:
I suppose so, it certainly looks like it. Will fix in the next revision.

K900 commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -43,14 +43,22 @@ let
 
   # Currently only enabling Rust by default on kernel 6.12+,
   # which actually has features that use Rust that we want.
   defaultRust = lib.versionAtLeast version "6.12" && rustAvailable;
```
We can drop the version check here too, and maybe this entire variable? Or maybe this should be `kernelSupportsRust && rustAvailable`? But then it won't be a default, it'll just be a `canWeActuallyBuildWithRust` or something. Any suggestions on cleaning this up?

timschumi replied:
I think I'll just drop the entire variable, but as a separate commit, in case someone wants to come back and extend the compatibility list for pre-6.12 again.

timschumi replied:
Thought about this while trying to implement it, wouldn't it make sense to have the "force Rust" flag affect the version check (i.e. `kernelSupportsRust`) instead of the `rustAvailable` flag? We are admittedly very lax with the exact version steps, so it would be more beneficial to the user to override that decision rather than "you don't have `rustc` installed".

K900 replied:
Hm, I thought we already did that?

timschumi replied:
Nope, the current logic is `(forceRust || defaultRust) && kernelSupportsRust;` (and with `kernelSupportsRust` only starting from 6.12 now, that is now a hard requirement for enabling rust and not forceable).

timschumi replied:
Done, and I tried making the surrounding bits a bit nicer as well.

timschumi committed:
linux/common-config: Overhaul Rust for Linux enabling logic

The table with kernel version and compiler requirements only starts from
Linux 6.12 now (the previous default), so update the force-enable logic
to accommodate users that don't want to follow those suggestions.

This includes printing warnings instead of asserting and allowing the
force-enable to override both the compiler check and the version (or
architecture) check.
